### ISSUE TITLE
Soft Delete Support (TypeORM)

### ISSUE THREAD

**Issue by exequiel09** [nestjsx/crud#433]
TypeORM recently release a new version containing the [Soft Delete functionality](https://github.com/typeorm/typeorm/pull/5034). I think we should also support this functionality.

**Comment by listochkin** [comment]
Here's how I did it for `User`:

**User Entity**
```TypeScript
  @DeleteDateColumn()
  deletedAt: Date;
```

**Users Service**
```TypeScript
  markDeleted(id) {
    this.usersRepository.softDelete({ id })
    return;
  }
```

**Users Controller**
```TypeScript
@Crud({
  ...
  query: {
    filter: {
      deletedAt: {
        $eq: null
      }
    },
  },
  ...
})
@Controller('users')
export class UsersController implements CrudController<User> {
  constructor(public service: UsersService) { }

  @Override()
  async deleteOne(@ParsedRequest() req: CrudRequest) {
    const id = req.parsed.paramsFilter
      .find(f => f.field === 'id' && f.operator === '$eq').value;
    const res = await this.service.markDeleted(id);
    return res;
  }
}
```

Pretty consize overall, but build-in support would still be nice.

**Comment by ruslanguns** [comment]
@listochkin good advice thanks!, I just have one question... I'd like to be able to list deletedUsers using CrudRequest feature for filtering, sorting or paging results from controller, but adding a custom request { withDeleted: true } , Could you please advice me on this?

```ts
const deletedUsers = await this.userRepository.find({ withDeleted: true, where: { deletedAt: Not(IsNull()) } });});
const deletedUsers = await this.userRepository.find(ids, { withDeleted: true, where: { deletedAt: Not(IsNull()) } });});
const deletedUser = await this.userRepository.findOne(id, { withDeleted: true, where: { deletedAt: Not(IsNull()) } });});
```

**Comment by listochkin** [comment]
@ruslanguns how about using filter for that? 

```JavaScript 
const deletedUsers = await fetch(`http://my.app/users?filter=${encodeURIComponent('deletedAt||$notnull')}`);
```

That would be an extra request on the UI (one for loading current users, another for deleted) but you could run them in parallel like this:

```
const [activeUsers, deletedUsers] = await Promise.all(
  fetch(`http://my.app/users`),
  fetch(`http://my.app/users?filter=${encodeURIComponent('deletedAt||$notnull')}`)
);

const allUsers = [...activeUsers, ...deletedUsers];
```

**Comment by JonatanSalas** [comment]
Would be pretty cool to support it via CRUD config like in this sample: 

```typescript
@Crud({
  model: {
    type: MyModel,
  },
  routes: {
    deleteOneBase: {
      softDelete: true,
    },
  },
})
@Controller('my-models')
export class MyModelController implements CrudController<MyModel> {
    constructor(public service: MyModelService) {}
}
```

**Comment by bestickley** [comment]
In the meantime, I created a helper function that you can user in your service:
```
// crudHelper.ts
export async function softDeleteOne<T>(
  this: TypeOrmCrudService<T>,
  req: CrudRequest
): Promise<void | T> {
  if (req.options.routes && req.options.routes.deleteOneBase) {
    const { returnDeleted } = req.options.routes.deleteOneBase;
    const found = await this.getOneOrFail(req, returnDeleted);
    const toReturn = returnDeleted
      ? plainToClass(this.entityType, { ...found })
      : undefined;
    await this.repo.softRemove(found);
    return toReturn;
  } else {
    throw new InternalServerErrorException("Incomplete CrudRequest");
  }
}
```
```
// user.service.ts
@Injectable()
export class UserService extends TypeOrmCrudService<User> {
  constructor(
    @InjectRepository(User) private readonly userRepository: Repository<User>
  )
  deleteOne(req: CrudRequest): Promise<void | User> {
  async deleteOne(req: CrudRequest): Promise<void | User> {
    return await softDeleteOne.call<
      TypeOrmCrudService<User>,
      [CrudRequest],
      Promise<void | User>
    >(this, req);
  }
  }
}
```

**Comment by quangtudng** [comment]
> @ruslanguns how about using filter for that?
> 
> ```js
> const deletedUsers = await fetch(`http://my.app/users?filter=${encodeURIComponent('deletedAt||$notnull')}`);
> ```
> 
> That would be an extra request on the UI (one for loading current users, another for deleted) but you could run them in parallel like this:
> 
> ```
> const [activeUsers, deletedUsers] = await Promise.all(
>   fetch(`http://my.app/users`),
>   fetch(`http://my.app/users?filter=${encodeURIComponent('deletedAt||$notnull')}`)
> );
> 
> const allUsers = [...activeUsers, ...deletedUsers];
> ```

The filter did not give me the deleted item unfortunately. Actually, i think it already ignored all the deleted item in the getMany route already

**Comment by Delta843** [comment]
Any updates on this? Would love to see this feature officially supported in the future

**Comment by danyalutsevich** [comment]
i guess thats what yall looking for
i dont know if that was already done in nestjsx/crud but this feature avaliable here https://github.com/gid-oss/dataui-nestjs-crud


<img width="327" alt="Screenshot 2024-03-11 at 17 21 29" src="https://github.com/nestjsx/crud/assets/83612717/6829d1f9-ae26-46f1-b8be-7c1693ec4db2">

**Comment by devdevdenis** [comment]
@danyalutsevich it works here as well. Tried it with: `@nestjsx/crud-typeorm: ^5.0.0-alpha.3`.
Most likely should be available in the latest tag: `v.5.0.1`
<img src="https://github.com/user-attachments/assets/a1e17d51-ef25-46e1-ad78-19c2c7f389e8" width="300">
